pcsx2/USB/usb-pad/usb-train.cpp

```diff
@@ -0,0 +1,546 @@
+		if (0x18 >= value)
+			return 0x23;
+		if (value >= 0xF8)
+			return 0xD7;
+		// We've trimmed 0x20 (0x8 for EB, 0x18 for release) leaving us with ~0xE0.
+		// 0xD7-0x23=0xB3 represents about 80% of the number space of 0xE0, so compress the remaining input values into that range.
+		u8 offset = 0x9 + value / 85;
+		return value / 5 * 4 + offset;
+	}
+
+#define get_ab(buttons) (button_at(buttons, CID_TC_A) | button_at(buttons, CID_TC_B))
+#define get_cd(buttons) (button_at(buttons, CID_TC_C) | button_at(buttons, CID_TC_D))
+#define swap_cd(buttons) ((button_at(buttons, CID_TC_C) << 1) | (button_at(buttons, CID_TC_D) >> 1))
+#define get_ss(buttons) (button_at(buttons, CID_TC_START) | button_at(buttons, CID_TC_SELECT))
+
+	// TrainControlID buttons are laid out in Type 2 ordering, no need to remap.
+	constexpr u8 dct01_buttons(u8 buttons) { return buttons; }
+	constexpr u8 dct02_buttons(u8 buttons)
+	{
+		return ((get_ab(buttons) << 2) | (swap_cd(buttons) >> 2) | get_ss(buttons));
+	}
+	constexpr u8 dct03_buttons(u8 buttons)
+	{
+		return (get_ab(buttons) | (button_at(buttons, CID_TC_CAMERA) >> 4) | ((get_cd(buttons) | get_ss(buttons)) << 1));
+	}
+
+	static void train_handle_data(USBDevice* dev, USBPacket* p)
+	{
+		TrainDeviceState* s = USB_CONTAINER_OF(dev, TrainDeviceState, dev);
+		TrainConOut out = {};
+
+		if (p->pid != USB_TOKEN_IN || p->ep->nr != 1)
+		{
+			Console.Error("Unhandled TrainController request pid=%d ep=%u", p->pid, p->ep->nr);
+			p->status = USB_RET_STALL;
+			return;
+		}
+
+		s->UpdateHatSwitch();
+		switch (s->type)
+		{
+			case TRAIN_TYPE2:
+				out.type2.control = 0x1;
+				out.type2.brake = s->passthrough ? s->data.brake : dct01_brake(s->data.brake);
+				out.type2.power = s->passthrough ? s->data.power : dct01_power(s->data.power);
+				out.type2.horn = 0xFF; // Button C doubles as horn.
+				out.type2.hat = s->data.hatswitch;
+				out.type2.buttons = dct01_buttons(s->data.buttons);
+				break;
+			case TRAIN_SHINKANSEN:
+				out.shinkansen.brake = s->passthrough ? s->data.brake : dct02_brake(s->data.brake);
+				out.shinkansen.power = s->passthrough ? s->data.power : dct02_power(s->data.power);
+				out.shinkansen.horn = 0xFF; // Button C doubles as horn, skip.
+				out.shinkansen.hat = s->data.hatswitch;
+				out.shinkansen.buttons = dct02_buttons(s->data.buttons);
+				break;
+			case TRAIN_RYOJOUHEN:
+				out.ryojouhen.brake = s->passthrough ? s->data.brake : dct03_brake(s->data.brake);
+				out.ryojouhen.power = s->passthrough ? s->data.power : dct03_power(s->data.power);
+				out.ryojouhen.horn = 0xFF; // Dedicated horn button, skip.
+				out.ryojouhen.hat = s->data.hatswitch & 0x0F;
+				out.ryojouhen.buttons = dct03_buttons(s->data.buttons);
+				break;
+			default:
+				Console.Error("Unhandled TrainController USB_TOKEN_IN pid=%d ep=%u type=%u", p->pid, p->ep->nr, s->type);
+				p->status = USB_RET_IOERROR;
+				return;
+		}
+
+		usb_packet_copy(p, &out, sizeof(out));
```

Comment:
Although wMaxPacketSize is equal to 8 bytes for all 3 controllers, 2 of them send only 6 bytes (?): 
https://marcriera.github.io/ddgo-controller-docs/controllers/usb/tcpp20009/#input
https://marcriera.github.io/ddgo-controller-docs/controllers/usb/tcpp20011/#input

I'd split TrainConOut in 3 distinct structs

Reply:
Makes sense. I found that even if sending 8 bytes, the other controllers seemed to work fine. But your suggestion is cleaner, I'll fix that up.